Override base implementations of WriteLineAsync in HttpResponseStreamWriter

david-acker opened the pull request:
## Description

Currently, `HttpResponseStreamWriter` uses the base implementation for `WriteLineAsync(string)` inherited from `TextWriter`. However, this base implementation apparently makes use of some synchronous operations. 

As a result, when attempting to write a long string (~20,000+ characters) to an `HttpResponse` body with `HttpResponseStreamWriter.WriteLineAsync(string)`, a `System.InvalidOperationException` will be thrown unexpectedly due to the synchronous operations used in the base implementation. This PR adds a fully asynchronous override for `WriteLineAsync(string)` to address this issue.

Fixes #38210

david-acker committed:
Override WriteLineAsync(string)

Tratcher commented:
So far so good. It looks like there are several overloads of WriteLineAsync that have a similar problem:
https://github.com/dotnet/runtime/blob/44387608aa85ed8ff5e08d3263ec14aab17c91f7/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L592-L604
https://github.com/dotnet/runtime/blob/44387608aa85ed8ff5e08d3263ec14aab17c91f7/src/libraries/System.Private.CoreLib/src/System/IO/TextWriter.cs#L639-L644

Can you test those as well?

david-acker commented:
Will do!

david-acker committed:
Override WriteLineAsync char array

david-acker committed:
Override WriteLineAsync char

Kahbazi commented on src/Http/WebUtilities/src/HttpResponseStreamWriter.cs:
```diff
@@ -419,6 +419,107 @@ private async Task WriteLineAsyncAwaited(ReadOnlyMemory<char> value)
+
+    private async Task WriteLineAsyncAwaited(char[] values, int index, int count)
+    {
+        await WriteLineAsync(values, index, count);
+        await WriteLineAsync(NewLine);
```
Is `WriteLineAsync` correct here!?
```suggestion
        await WriteAsync(values, index, count);
        await WriteAsync(NewLine);
```

david-acker replied:
Ah, yeah that's incorrect. It should be `WriteAsync`.

Kahbazi commented on src/Http/WebUtilities/src/HttpResponseStreamWriter.cs:
```diff
@@ -419,6 +419,107 @@ private async Task WriteLineAsyncAwaited(ReadOnlyMemory<char> value)
+
+    private async Task WriteLineAsyncAwaited(char[] values, int index, int count)
+    {
+        await WriteLineAsync(values, index, count);
+        await WriteLineAsync(NewLine);
```
Is `WriteLineAsync` correct here!?
```suggestion
        await WriteAsync(values, index, count);
        await WriteAsync(NewLine);
```

david-acker committed:
Update src/Http/WebUtilities/src/HttpResponseStreamWriter.cs

Fix method call

Co-authored-by: Kahbazi <akahbazi@gmail.com>

Tratcher commented:
Interesting build error. `src/Http/WebUtilities/src/HttpResponseStreamWriter.cs(385,26): error RS0027: (NETCORE_ENGINEERING_TELEMETRY=Build) Symbol 'WriteLineAsync' violates the backcompat requirement: 'Public API with optional parameter(s) should have the most parameters amongst its public overloads'. See 'https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md' for details.`

Since it's an override, I think that should be suppressed like WriteAsync does:
https://github.com/dotnet/aspnetcore/blob/0ff49cbe9f7caa5960f4e1d5abc39b50855c1cdd/src/Http/WebUtilities/src/HttpResponseStreamWriter.cs#L333-L334

david-acker committed:
Suppress RS0027

david-acker committed:
Fix method call

david-acker committed:
Use correct overload for tests, add cases

Tratcher committed:
Suppress ApiDesign RS0027

david-acker committed:
Add test for count argument of zero

david-acker committed:
Fix merge conflict

david-acker commented:
Fixed one spot where `WriteAsyncAwaited` was being called instead of `WriteLineAsyncAwaited` and added index and count arguments the char array overload tests so the correct overload was used.
